Add validation for class_weight keys in standardize_sample_or_class_weight.

Aaraviitkgp opened the pull request:
issue #98283
Invalid class_weight silently ignored in model.fit(), no error raised

1. tensorflow/python/keras/engine/training_utils_v1.py
Modified standardize_sample_or_class_weights() function (lines ~693-720, ~745-755)
Added validation logic for both single-output and multi-output models
Validates class_weight keys while preserving sample_weight functionality
Provides detailed error messages with key type information
2. tensorflow/python/keras/engine/data_adapter.py
Enhanced _make_class_weight_map_fn() function (lines ~1412-1450)
Added validation at the data adapter level for additional safety
Added automatic conversion of valid string keys to integers
Provides consistent error messages across validation points
3. tensorflow/python/keras/engine/training_utils_v1_test.py (NEW FILE)
Comprehensive test suite covering all validation scenarios
Tests for invalid keys (strings, floats) that should be rejected
Tests for valid keys (integers, output names, string numbers) that should be accepted
Tests for multi-output model validation
Ensures sample_weight validation remains unaffected

Aaraviitkgp committed:
Add validation for class_weight keys in standardize_sample_or_class_weights

mihaimaruseac reviewed (changes requested):
Missing BUILD file, so the tests were not run. Please make sure to run the tests before submitting the PR.

Also, please eliminate changes not related to the PR. It is your responsibility to clean after the AI if you use AI to create the PR.

Aaraviitkgp commented:
@mihaimaruseac sorry for the inconvenience, I will do the changes as fast as possible.

Aaraviitkgp committed:
build file added

Aaraviitkgp commented:
@mihaimaruseac I have added the build file and edited out the pr and once again sorry for the inconvenience.

Aaraviitkgp committed:
did some more changes asked

Aaraviitkgp committed:
made all changes asked

Aaraviitkgp committed:
pylint issue handled

mihaimaruseac commented on tensorflow/python/keras/engine/training_utils_v1_test.py:
```diff
@@ -0,0 +1,91 @@
+
+
+if __name__ == '__main__':
+  tf.test.main()
```
Please configure your editor to insert newline at the end of all files. I already gave a review feedback on this vein in this PR.

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,7 +1412,30 @@ def _make_class_weight_map_fn(class_weight):
   """
-  class_ids = list(sorted(class_weight.keys()))
+  # Validate that all keys are integers
+  for key in class_weight.keys():
+    if not isinstance(key, (int, numpy_compat.integer_types)):
+      try:
+        # Try to convert to int if it's a string representation
+        if isinstance(key, str):
+          int(key)
+        else:
+          raise ValueError()
+      except (ValueError, TypeError):
+        raise ValueError(f"Invalid class_weight key: '{key}'. "
+                        f"Class weight keys must be integers representing "
+                        f"class indices,"
+                        f"got key of type {type(key).__name__}.")
```
Why not just convert directly to the key used in the `converted_class_weight` below? Right now you are parsing the string twice.

Aaraviitkgp committed:
all changes made

Aaraviitkgp commented:
@mihaimaruseac made all the changes please check.

Aaraviitkgp committed:
made changes

Aaraviitkgp commented:
@mihaimaruseac please check, I have made all the changes that you asked for.

Aaraviitkgp commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,17 +1412,51 @@ def _make_class_weight_map_fn(class_weight):
-        "than the number of classes, found {}").format(class_weight)
-    raise ValueError(error_msg)
-
+    # Check if any class weight values are complex tensors
+    has_complex_weights = False
+    for value in converted_class_weight.values():
+      try:
+        # Convert to tensor and check if complex
+        tensor_val = tensor_conversion.convert_to_tensor_v2_with_dispatch(value)
+        if tensor_val.dtype.is_complex:
+          has_complex_weights = True
+          break
+      except:
+        pass
```
its just a extra safety net.

Aaraviitkgp committed:
removed complex check

Aaraviitkgp committed:
removed line

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,17 +1412,51 @@ def _make_class_weight_map_fn(class_weight):
         "than the number of classes, found {}").format(class_weight)
-    raise ValueError(error_msg)
-
+    raise ValueError(error_msg) 
```
You should keep the empty line. Just don't add trailing whitespace on it.

Aaraviitkgp replied:
done

Aaraviitkgp committed:
new changes made

Aaraviitkgp commented:
@mihaimaruseac please check all the changes have been made.

Aaraviitkgp committed:
all changes made

Aaraviitkgp commented:
@mihaimaruseac please check the changes, I have added a new test file for data adapter tests ,added it in build file also removed the data adapter tests from training utils v1 test file, removed the complex part and replaced type(0) with numpy_compat.integer_types.

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,17 +1412,52 @@ def _make_class_weight_map_fn(class_weight):
+      tensor_val = tensor_conversion.convert_to_tensor_v2_with_dispatch(value)
+      if tensor_val.dtype.is_complex:
+        raise ValueError(
+            f"Complex class weights are not supported. "
```
Wait, I see this now. Why are complex values not supported?

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,17 +1412,52 @@ def _make_class_weight_map_fn(class_weight):
         "than the number of classes, found {}").format(class_weight)
     raise ValueError(error_msg)
-
+   
```
You still have trailing whitespace. Please configure your editor to strip in on save, I have to mention this in almost every round of review.

Aaraviitkgp committed:
complex bit removed and trailing whitespaces removed

Aaraviitkgp committed:
pylint error resolved

Aaraviitkgp committed:
fixed changes

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,7 +1412,28 @@ def _make_class_weight_map_fn(class_weight):
     weighting.
   """
-  class_ids = list(sorted(class_weight.keys()))
+  # Convert string keys to integers
+  converted_class_weight = {}
+  for key, value in class_weight.items():    # Convert keys to integers
```
Please remove the comment in both lines, it's useless.

Please make sure to always clean after the AI if using an LLM to generate the code.

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,7 +1412,28 @@ def _make_class_weight_map_fn(class_weight):
+      converted_class_weight[key] = value
+    elif isinstance(key, str):
+      try:
+        int_key = int(key)
+        converted_class_weight[int_key] = value
```
These two lines can be combine into a single one

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1421,9 +1442,10 @@ def _make_class_weight_map_fn(class_weight):
 
   class_weight_tensor = tensor_conversion.convert_to_tensor_v2_with_dispatch(
-      [class_weight[int(c)] for c in class_ids]
+      [converted_class_weight[int(c)] for c in class_ids]
```
Do we still need the `int` conversion here?

mihaimaruseac commented:
> made the logic same

It still doesn't look the same to me. Please justify any differences or make it exactly the same.

Aaraviitkgp committed:
fixed changes

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter.py:
```diff
@@ -1412,7 +1412,26 @@ def _make_class_weight_map_fn(class_weight):
+                      f"class indices, "
+                      f"got key of type {type(key).__name__}.")
+
+  # Validate class indices are contiguous starting from 0
```
I suggest removing the comment.

Aaraviitkgp replied:
Well then I will hardcode it so that it is same.I will update it by today.

Aaraviitkgp committed:
fixed more changes

mihaimaruseac commented on tensorflow/python/keras/engine/training_utils_v1.py:
```diff
@@ -685,36 +685,86 @@ def standardize_sample_or_class_weights(x_weight, output_names, weight_type):
   """
   if x_weight is None or (isinstance(x_weight, (list, tuple)) and
-                          len(x_weight) == 0):  # pylint: disable=g-explicit-length-test
+                          len(x_weight) == 0):
```
I'm not sure the pylint warning here can be removed

mihaimaruseac commented on tensorflow/python/keras/engine/training_utils_v1.py:
```diff
@@ -685,36 +685,86 @@ def standardize_sample_or_class_weights(x_weight, output_names, weight_type):
-  if isinstance(x_weight, collections.abc.Mapping):
-    generic_utils.check_for_unexpected_keys(weight_type, x_weight, output_names)
-    x_weights = []
+  if isinstance(x_weight, dict):
```
Why did this change from the more generic type?

Aaraviitkgp replied:
@mihaimaruseac I am replacing if isinstance(x_weight, dict): with if isinstance(x_weight, collections.abc.Mapping): this will it make more generic while keeping it same like data_handler.

mihaimaruseac replied:
yes please

Aaraviitkgp committed:
made all changes

mihaimaruseac committed:
Add missing space

mihaimaruseac commented:
I'll check the internal failure on Monday when I can do that.

Aaraviitkgp commented:
@mihaimaruseac did you see the internal logs where the error is coming.

mihaimaruseac commented:
Since this is keras, it should use the native Bazel rules, not the TF variants in BUILD files:

```
Prefer strict rule variants, e.g. py_strict_binary, py_strict_library, py_strict_test, tf_py_strict_test, cuda_py_strict_test, etc.
```

Aaraviitkgp committed:
removed tf from build file

mihaimaruseac commented on tensorflow/python/keras/engine/BUILD:
```diff
@@ -1,8 +1,7 @@
 
-load("//tensorflow:py.default.bzl", "py_library")
-
+load("//tensorflow:py.default.bzl", "py_library","tf_py_test")
```
This needs to change too

Aaraviitkgp committed:
made changes asked

mihaimaruseac commented on tensorflow/python/keras/engine/data_adapter_test.py:
```diff
@@ -0,0 +1,108 @@
+# ==============================================================================
+"""Tests for tensorflow.python.keras.engine.data_adapter."""
+
+import tensorflow as tf
```
Oh, we cannot have this hourglass import. It basically needs to load the entirety of TensorFlow in order to run the tests.

You should be able to only import the proper internal paths instead (in both test files)

Aaraviitkgp committed:
made changes asked

Aaraviitkgp committed:
pylint error resolved

Aaraviitkgp committed:
made changes

Aaraviitkgp committed:
pylint error resolved

Aaraviitkgp commented:
@mihaimaruseac please check it.

Aaraviitkgp commented:
@mihaimaruseac what failed now please check. because it worked perfectly when I tested.

mihaimaruseac commented:
```
==================== Test output for //`tensorflow/python/keras/engine:training_utils_v1_test:
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/training_utils_v1_test.runfiles/org_tensorflow/tensorflow/python/keras/engine/training_utils_v1_test.py", line 17, in <module>
    from tensorflow.python.keras.engine import training_utils_v1
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/training_utils_v1_test.runfiles/org_tensorflow/tensorflow/python/keras/engine/training_utils_v1.py", line 28, in <module>
    from tensorflow.python import tf2
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/training_utils_v1_test.runfiles/org_tensorflow/tensorflow/python/tf2.py", line 21, in <module>
    from tensorflow.python.platform import _pywrap_tf2
ImportError: cannot import name '_pywrap_tf2' from 'tensorflow.python.platform' (/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/training_utils_v1_test.runfiles/org_tensorflow/tensorflow/python/platform/__init__.py)

==================== Test output for //tensorflow/python/keras/engine:data_adapter_test:
Traceback (most recent call last):
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/data_adapter_test.runfiles/org_tensorflow/tensorflow/python/keras/engine/data_adapter_test.py", line 17, in <module>
    from tensorflow.python.framework import dtypes
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/data_adapter_test.runfiles/org_tensorflow/tensorflow/python/framework/dtypes.py", line 28, in <module>
    from tensorflow.python import pywrap_tensorflow  # pylint: disable=unused-import
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/data_adapter_test.runfiles/org_tensorflow/tensorflow/python/pywrap_tensorflow.py", line 37, in <module>
    self_check.preload_check()
  File "/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/data_adapter_test.runfiles/org_tensorflow/tensorflow/python/platform/self_check.py", line 63, in preload_check
    from tensorflow.python.platform import _pywrap_cpu_feature_guard
ImportError: cannot import name '_pywrap_cpu_feature_guard' from 'tensorflow.python.platform' (/b/f/w/bazel-out/k8-opt/bin/tensorflow/python/keras/engine/data_adapter_test.runfiles/org_tensorflow/tensorflow/python/platform/__init__.py)
================================================================================
```

Aaraviitkgp commented:
@mihaimaruseac I am not able to understand where the error is coming from , please help.

Aaraviitkgp commented:
@mihaimaruseac I am not able to resolve this ImportError: cannot import name '_pywrap_tf2' from 'tensorflow.python.platform'.Can you suggest any viable method to resolve this error.
